api/conversion/machine_test.go

```diff
@@ -0,0 +1,461 @@
+          caCertPath: "/tmp/wonderbolt.sock"
+          discovery:
+            bootstrapToken:
+              token: "r41nb0w"
+              apiServerEndpoint: "http://friendship.castle"
+              caCertHashes:
+              - "c313571a"
+              unsafeSkipCAVerification: false
+            fileDiscovery:
+              kubeConfig: "/tmp/rainbow.conf"
+            tlsBootstrapToken: "7w1l1ght"
+            timeout: "8791ms"
+          controlPlane:
+            localAPIEndpoint:
+              advertiseAddress: "friendship.castle"
+              bindPort: 1235
+      additionalUserDataFiles:
+        - path: "/tmp/tank"
+          owner: "rd:rd"
+          permissions: "0640"
+          content: "the best pet for me"
+`
+
+func getMachine(t *testing.T) (*capiv1a1.Machine, *capav1a1.AWSMachineProviderSpec) {
+	scheme := runtime.NewScheme()
+	capiv1a1.SchemeBuilder.AddToScheme(scheme)
+	capav1a1.SchemeBuilder.AddToScheme(scheme)
+
+	decoder := serializer.NewCodecFactory(scheme).UniversalDecoder()
+
+	var (
+		machine    capiv1a1.Machine
+		awsMachine capav1a1.AWSMachineProviderSpec
+	)
+
+	if _, _, err := decoder.Decode([]byte(exampleMachineYAML), nil, &machine); err != nil {
+		t.Fatalf("failed to decode example: %v", err)
+	}
+
+	if machine.Spec.ProviderSpec.Value == nil {
+		t.Fatalf("No providerspec found")
+	}
+
+	if _, _, err := decoder.Decode(machine.Spec.ProviderSpec.Value.Raw, nil, &awsMachine); err != nil {
+		t.Fatalf("failed to decode example providerSpec: %v", err)
+	}
+
+	return &machine, &awsMachine
+}
+
+func TestConvertMachine(t *testing.T) {
+	var (
+		newMachine       capiv1a2.Machine
+		newAWSMachine    capav1a2.AWSMachine
+		newKubeadmConfig cabpkv1a2.KubeadmConfig
+	)
+
+	oldMachine, oldAWSMachine := getMachine(t)
+	oldCluster, oldAWSCluster := getCluster(t)
+
+	converter := NewMachineConverter(oldCluster, oldMachine)
+
+	if err := converter.GetMachine(&newMachine); err != nil {
+		t.Fatalf("Unexpected error converting machine: %v", err)
+	}
+
+	assert := asserter{t}
+
+	if oldMachine == nil {
+		t.Fatalf("Unexpectedly nil machine")
+	}
+
+	assert.stringEqual(oldMachine.ObjectMeta.Name, newMachine.ObjectMeta.Name, "machine name")
+	assert.stringEqual(oldMachine.ObjectMeta.Namespace, newMachine.ObjectMeta.Namespace, "machine namespace")
+
+	assert.stringEqual(oldMachine.Spec.ObjectMeta.Name, newMachine.Spec.ObjectMeta.Name, "node name")
+	assert.stringEqual(oldMachine.Spec.ObjectMeta.Namespace, newMachine.Spec.ObjectMeta.Namespace, "node namespace")
+
+	assert.stringPtrEqual(&oldMachine.Spec.Versions.ControlPlane, newMachine.Spec.Version, "version")
+
+	assert.stringPtrEqual(oldMachine.Spec.ProviderID, newMachine.Spec.ProviderID, "provider ID")
+
+	if err := converter.GetAWSMachine(&newAWSMachine); err != nil {
+		t.Fatalf("Unexpected error converting AWSMachine: %v", err)
+	}
+
+	t.Logf("converted machine: %+v", newAWSMachine)
+
+	// Pull the provider ID from the machine (why is it in both places?)
```

Comment:
It's in both places because the infra provider is the primary component responsible for the value, but CAPI needs it for setting node refs, so CAPI copies it from the infra resource to the machine.